Allow removal of disconnected Jetpack purchases

chrismccluskey opened the pull request:
Due the UI preventing any management of purchases if the user has no sites, an edge case
happens where a user who has no sites, made a Jetpack purchase and then disconnected
that Jetpack purchase is unable to manage their purchase. This will move the check that
sites exist to after the call to fetch user purchases, allowing all purchases to be shown,
regardless of the number of sites to which a user belongs. For listing purchases, if there
are no purchases found, it will show the no sites message. Managing purchases have the
check for user sites removed entirely as manage purchases will redirect to /me/purchases if
something is wrong.

#### Changes proposed in this Pull Request

* For `/me/purchases` move the `userHasNoSites` check from the controller to inside the `PurchasesList` component after no purchases have been found.
* For `/me/purchases/:site/:purchaseId` remove the check for `userHasNoSites` entirely. The component will redirect to `/me/purchases` if an invalid combination of `:site` and `:purchaseId` are encountered.

#### Testing instructions



* On a brand new user with no existing sites, purchase a Jetpack subscription
* Disconnect the purchased Jetpack subscription
* View this branch and navigate to `/me/purchase` and if everything works properly you should be able to remove your Jetpack subscription (if you are outside of the refund window) or cancel & refund your Jetpack subscription (if you are within the refund window)

sirreal commented:
Cross-referencing: p4jECJ-1qA-p2

sirreal commented on client/me/purchases/purchases-list/index.jsx:
```diff
@@ -30,6 +30,11 @@ import {
+import NoSitesMessage from 'components/empty-content/no-sites-message';
+
+const user = userFactory();
+const userHasNoSites = () => user.get().site_count <= 0;
```
Rather pulling in lib/user here, let's connect to our redux store with this selector:

https://github.com/Automattic/wp-calypso/blob/3a0e026843cca5d4fa0a1ea14ccb371765bd6f99/client/state/current-user/selectors.js#L81-L94

sirreal commented on client/me/purchases/purchases-list/index.jsx:
```diff
@@ -69,6 +74,15 @@ class PurchasesList extends Component {
+				return (
+					<Main>
+						<PageViewTracker path="/me/purchases" title="Purchases > No Sites" />
+						<PurchasesHeader section={ 'purchases' } />
```
```suggestion
						<PurchasesHeader section="purchases" />
```

chrismccluskey commented:
Thank you for the reviews @sirreal and @DavidRothstein ! I'm working on these changes now.

> Is there a different way that you got the link to appear in your testing?

I think in this case the time between my creating my test account to reproduce this issue and the time to develop a fix might have meant the purchase fell outside of the refund window and thus the only option I was presented with was just to Remove. I've created a new test account and can reproduce this - I don't have any link to remove, my screen looks like what you've screen shotted.

I'll work to address this issue and update this PR.

chrismccluskey commented:
I believe I've found and resolved the issue @DavidRothstein originally brought up, which turned out to have me following in the footsteps of @sirreal when he was fixing this issue originally, enabling similar pathways through the code to actually let disconnected Jetpack users cancel and refund their purchase. These changes are now ready for review again, and if either of you could spare a few moments to look over my work that would be greatly appreciated. Along with that, there are a few key things I want to bring to your attention about my recent changes in this PR:

- I've [updated the `getSelectedSite` function to return `false` when no site is found](https://github.com/Automattic/wp-calypso/pull/37542/files#diff-1737c4fc051ae98192e462012fbb5a85), instead of `null`. I made this change because I found multiple cases of `propTypes` allowing either a `bool` or an `object` for the `selectedSite` prop, leading me to think that a value of `false` is actually expected here. I was otherwise getting propType warnings in console.
- The [`ManagePurchase` component now requires a `siteSlug` instead of a `site`](https://github.com/Automattic/wp-calypso/pull/37542/files#diff-e2b8265ef22dd8af459ffe058b1b3b08)

I don't think these changes will cause problems, and in my testing didn't appear to cause problems, but I thought it's worth mentioning as they might be bigger changes than I realize.

chrismccluskey commented:
One last thing I forgot to mention: You won't be able to actually cancel the purchase unless you also apply D35694-code
Without the patch applied, you should be able to get as far as making the request to the backend to cancel, which will fail.

chrismccluskey commented:
> I would strongly discourage this. Changing the behavior of `getSelectedSite` may have implications we don't understand and can't anticipate.

Understood and reverted. I've also reverted the changes I've made to other `propType` declarations elsewhere. Finally, to remove the propType warnings on the component that was generating them, I've just removed the `isRequired` restriction to allow null values for the `selectedSite` prop on the `CancelPurchaseButton` component.

I greatly appreciate your comments on how `propTypes` are used as I'm new to using them myself.

> It's likely beyond the scope of your work here, but we can now write code in Calypso with TypeScript. TypeScript provides much stronger confidence in application correctness in addition to improved developer experience thanks to things like much better completion suggestions in the editor.

This is great to know. I've never touched TypeScript, so I think it does fall beyond the scope of this task, but I'll start learning it so I can make better calypso contributions in the future. 👍 

---

That said, with the change @sirreal discouraged reverted (along with other checks to accompany that change reverted), this code is ready for review again. Thank you both for your help with this and it is my hope this will be the final round of review for this issue.

chrismccluskey commented:
Also, I'm noticing these e2e tests are not passing. Do I need to do anything to make these pass? The errors in the details don't make sense to me.

sirreal commented:
> Also, I'm noticing these e2e tests are not passing. Do I need to do anything to make these pass? The errors in the details don't make sense to me.

There were some issues with e2e recently. Often rebasing or simply rerunning them (remove and re-add the "needs review" label) will help.

sirreal commented on client/me/purchases/cancel-purchase/button.jsx:
```diff
@@ -34,7 +34,7 @@ import { getDowngradePlanFromPurchase } from 'state/purchases/selectors';
 	static propTypes = {
 		purchase: PropTypes.object.isRequired,
-		selectedSite: PropTypes.object.isRequired,
+		selectedSite: PropTypes.object,
```
If you observed a `null` selectedSite here, I'd suggest auditing the entire file. For example, this is potentially problematic:

https://github.com/Automattic/wp-calypso/blob/0753a38fc4c61df18e8450abe67b84ad87c7511d/client/me/purchases/cancel-purchase/button.jsx#L79

```js
const { slug } = null;
// Throws:
// TypeError: Cannot destructure property `slug` of 'undefined' or 'null'.
```

chrismccluskey replied:
Do you think it would be better to keep `selectedSite` as a required prop and allow the `propType` warnings to persist? The idea being that in the rare case where a user with no sites and disconnected Jetpack purchases wants to cancel their purchase, we're ok with a warning in the console or otherwise defying this propType.

I'm operating under the assumption that warnings are something I want to avoid introducing but in this case I'm wondering what my alternatives are. I've had to remove the `siteSelection` middleware from the cancel purchase route because customers in this scenario won't have any sites to select from as it was preventing the route from being hit by displaying a "no sites" message. And so at this point in the code, I don't have a site to provide this component with.

As you mention, these warnings are weak, and I was able to cancel the purchase just fine despite getting the warning in my console. I'm wondering if that's the right thing to do. Otherwise, my only other idea would be to pass on the fact that we're dealing with a purchase without a site by passing a custom (hardcoded) object that meets the `propType` requirement, but I'm not sure I like that approach.

sirreal replied:
> Do you think it would be better to keep `selectedSite` as a required prop and allow the `propType` warnings to persist? The idea being that in the rare case where a user with no sites and disconnected Jetpack purchases wants to cancel their purchase, we're ok with a warning in the console or otherwise defying this propType.

`propTypes` describe what we expect to find in the component. We should not see any `propTypes` warnings, they indicate a component is operating under abnormal conditions, or that our assumptions about the runtime behavior of the application are inaccurate. These are both problems and should concern us. The `propTypes` as currently written seem accurate.

From what I've understood, `selectedSite` may be `null`. You've observed `propTypes` warnings indicating that. At this point, we should have strong confidence that this component will behave correctly under all of the normal conditions:
- `this.props.selectedSite = null`
- `this.props.selectedSite = { /* …siteObject */ }`

At a glance, and I have not looked closely at the component, I would expect this line to error when `thes.props.selectedSite` is `null`:

https://github.com/Automattic/wp-calypso/blob/0753a38fc4c61df18e8450abe67b84ad87c7511d/client/me/purchases/cancel-purchase/button.jsx#L79

If that line is indeed reachable with a `null` `selectedSite`, additional changes should to be made to protect against runtime errors in the application.

> […] I was able to cancel the purchase just fine despite getting the warning in my console.

I think it's easy to see the line I've reference would error if it is reached with a `null` `selectedSite`. Did you attempt to hit that code path? Are you confident it is unreachable?

> […] Otherwise, my only other idea would be to pass on the fact that we're dealing with a purchase without a site by passing a custom (hardcoded) object that meets the `propType` requirement, but I'm not sure I like that approach.

It may be a technically feasible solution, but I don't like that approach either 🙂 Fighting complexity with complexity is a losing long-term strategy.

DavidRothstein replied:
It looks to me like there are a few places, not just limited to this file, that won't work well if the selected site isn't passed in (not all will be fatals like the above but they will all have consequences).  The ones I found:
- `CancelPurchaseLoadingPlaceholder`
- `CancelPurchase::redirect()`
- `ProductLink`
- `CancelPurchaseForm`

As an example of the effect, you can see that if you go to cancel a purchase with this PR (doesn't have to be a disconnected Jetpack plan; can be a WordPress.com plan or something else) the "View Plan Features" link is missing on the cancellation screen.  That comes from the `ProductLink` component.

It seems like all of the above places don't actually require a site object; they really just require the site slug (and in one case, `CancelPurchaseForm`, the site ID)?  But we already have the site slug, and we can get the site ID from the purchase object (the same way the `ManagePurchase` component already gets it).  So if the components are refactored to expect those properties only, rather than a site object, you could probably make it work that way.  (Not sure about `CancelPurchaseForm` though; that uses the site ID in a tricky way involving an API call via `QuerySitePlans`, and I'm not sure what the consequences are of having that API call be rejected as unauthorized since the user isn't part of the site.)

chrismccluskey replied:
Thank you both for your help with this.

@sirreal asked:
> Did you attempt to hit that code path? Are you confident it is unreachable?

I am somewhat confident this code path is not reachable, as it only appears if the purchase being cancelled is a domain registration, and in these cases we are cancelling a Jetpack registration. I didn't attempt to hit this code path because I'm not sure how I could create an account with a domain registration that is not associated with a site to even reach that code path.

@DavidRothstein suggested:
> It seems like all of the above places don't actually require a site object; they really just require the site slug (and in one case, CancelPurchaseForm, the site ID)? [...] So if the components are refactored to expect those properties only, rather than a site object, you could probably make it work that way.

This makes sense, and it sounds like a good way forward. This way I can keep the propTypes accurate and only require the string that I need, instead of the entire object. On the topic of `QuerySitePlans` I'm a little unsure what this syntax does:

https://github.com/Automattic/wp-calypso/blob/0753a38fc4c61df18e8450abe67b84ad87c7511d/client/components/marketing-survey/cancel-purchase-form/index.jsx#L752

I would interpret that as "If `selectedSite` is present, then include `QuerySitePlans` component". If I'm understanding that correctly, I wouldn't expect this `render` function to generate an error, but it also wouldn't include the component. 

I'd like to include the component, but I'm thinking that would require changing the way we handle disconnected Jetpack purchases by not removing the customer as a user from that site - which might have other side effects. I'll test this but I'm guessing if the correct ID were to be passed to `QuerySitePlans`, the backend would not return a match. If the backend does return a match, perhaps the solution to this issue can be handled entirely on the backend, by returning sites the user is not a member on, but has purchased a Jetpack plan for.

DavidRothstein commented on client/me/purchases/purchases-list/index.jsx:
```diff
@@ -69,6 +70,15 @@ class PurchasesList extends Component {
 		}
 
 		if ( this.props.hasLoadedUserPurchasesFromServer && ! this.props.purchases.length ) {
+			if ( this.props.hasMoreThanOneSite ) {
```
Aren't you checking the wrong thing here?  I think we actually want this to check if the user has no sites, not if the user has more than one site.

chrismccluskey replied:
🤦‍♂️ yes, I am. Thank you. I'll fix that.

DavidRothstein commented on client/me/purchases/manage-purchase/index.jsx:
```diff
@@ -230,7 +230,7 @@ class ManagePurchase extends Component {
 		const { id } = purchase;
 
-		if ( ! isCancelable( purchase ) || isPartnerPurchase( purchase ) || ! this.props.site ) {
+		if ( ! isCancelable( purchase ) || isPartnerPurchase( purchase ) || ! this.props.siteSlug ) {
```
Note sure if the `this.props.siteSlug` check is required here, since that's already supposed to be a required property.  I guess it doesn't hurt though :smile:

chrismccluskey replied:
Oh, good catch. I guess I'm not sure what's better here: leaving it out and just letting propTypes warn us when it fails, or keeping it in, even though it shouldn't ever be encountered if propTypes is preventing us from introducing those edge cases. Since I put it in during this PR, I'll just take it out, unless the extra defensiveness is encouraged :)

sirreal commented:
@DavidRothstein brings up a good point https://github.com/Automattic/wp-calypso/pull/37542#discussion_r349707471. By removing the siteSelection middleware, the selectedSite and related properties may become `null` anywhere in the tree. We need to make sure the entire component tree works well in the absence of a site.

One method of identifying issues is to change the `selectedSite` in `connect` to always return `null`, then use the application and chase errors.

chrismccluskey commented:
> One method of identifying issues is to change the `selectedSite` in `connect` to always return `null`, then use the application and chase errors.

That's a great idea! I'll give that a try, and try to address those concerns one by one by refactoring components into just the information they need if possible.

chrismccluskey commented:
It looks like everything in the `CancelPurchaseForm` component tree can be refactored to use `siteSlug` and `purchases.siteId` with one exception: `CancelPurchaseForm` needs to pass along `selectedSite` in order for `enrichedSurveyData` to use the `selectedSite.options.created_at` value. This data is sourced from the list of sites obtained from the backend - which will never include a disconnected Jetpack purchase.

There are a few ways I can think of refactoring this, and I'm unsure what's best:
1. Make the `/me/purchases` API endpoint return the site created date along with the `siteId` that is provided for each purchase. This might require a database lookup on the API endpoint. Refactor all calls to `enrichedSurveyData` to remove the `site` argument (altering the function signature).
2. Allow the `site` argument on `enrichedSurveyData` to be null, omitting the `daysSinceSiteCreation` metric for disconnected Jetpack purchases (which might be OK here - is this even relevant for Jetpack purchases?)
3. Don't send the marketing survey (don't even pop it up) if this is a cancellation of a Jetpack (or disconnected Jetpack) purchase.

I think if left to my own to decide, I like the first option 1 the most. Still, I'm unsure which is _best_. Do either of you have an opinion on how to handle this?

chrismccluskey commented:
As for the `ProductLink` component, that is easily made visible with the refactoring to use `siteSlug` instead of `selectedSite.slug`, but clicking the link to view the plan takes you to a route that has the `siteSelection` middleware, preventing that page from being shown. I can refactor this route to not use `siteSelection` or we could opt to not display the `ProductLink` component if there isn't a matching site to go with it. I'd also like to hear if either of you have an opinion on what to do here as well, if you don't mind. This is turning out to be quite a long PR so perhaps if refactoring this route is the right choice it's better to do it with a separate PR?

DavidRothstein commented:
Ah, I missed both those things.... Yeah, this is more complicated than expected.

I think either your option 1 or 2 makes sense for the `enrichedSurveyData` case.

And for `ProductLink`, I agree that it's OK if we skip showing the "View Plan Features" link in the case of a disconnected Jetpack site.  (Just as long as we still show it for all other cases.)

chrismccluskey commented:
Alright, I think this PR is finally in a place where it is ready for what I do hope is at least close to a final review 😄. Thank both you for your help. In this latest round of commits, I've done the following:

- [Removed the `selectedSite` requirement](https://github.com/Automattic/wp-calypso/pull/37542/commits/7b09fb0d43fb345bbca2ddc7ddadc96619d8beb4) where it was required in the relevant components that take you along the "Cancel Subscription and Refund" code path. I was able to reuse some existing properties like `purchases.siteId`, `site` and `siteSlug` to avoid passing `selectedSite` through to these components. 
- [The `ProductLink` component now only shows if the selected site is not null](https://github.com/Automattic/wp-calypso/pull/37542/commits/928966549f86dbb01f716a0fb8cb76e26facadf1). I was able to view this link on other Jetpack and wpcom plans.
- [Added `blogCreatedDate` property to `purchase` objects](https://github.com/Automattic/wp-calypso/pull/37542/commits/57e6d106e3b1d7f3ec71b266ad2614cacbdd07ce#diff-4430bf8471f224634878b049437f0a60R39). This requires D35938-code.
- [Refactored `enrichedSurveyData`](https://github.com/Automattic/wp-calypso/pull/37542/commits/57e6d106e3b1d7f3ec71b266ad2614cacbdd07ce), changing its function signature to remove the `site` argument, updated all instances that call this function. The `site` argument was just supplying the `site.options.created_at` date which is now available from the `purchase` argument.

As I mentioned, this requires D35938-code. There was another patch mentioned in this PR but it has since been merged.

DavidRothstein commented on client/me/purchases/cancel-purchase/index.jsx:
```diff
@@ -213,7 +206,7 @@ class CancelPurchase extends React.Component {
 					<div className="cancel-purchase__purchase-name">{ purchaseName }</div>
 					<div className="cancel-purchase__site-title">{ siteName || siteDomain }</div>
-					<ProductLink purchase={ purchase } selectedSite={ this.props.selectedSite } />
+					<ProductLink purchase={ purchase } siteSlug={ this.props.siteSlug } />
```
This change looks wrong -- the component still expects `selectedSite`.  So this seems to still be causing the "View Plan Features" link to disappear on the cancellation form, for all products including WordPress.com.

chrismccluskey replied:
Reverted in 1e6c160

DavidRothstein commented on client/me/purchases/cancel-purchase/button.jsx:
```diff
@@ -299,7 +298,7 @@ class CancelPurchaseButton extends Component {
 					onInputChange={ this.onSurveyChange }
 					purchase={ purchase }
-					selectedSite={ selectedSite }
+					selectedSite={ site }
```
I don't think the `site` property actually exists?

This means we're passing a null value to the cancellation form (for all products including WordPress.com), which in turn means that the `selectedSite && <QuerySitePlans siteId={ selectedSite.ID } />` in that component (discussed previously) never runs.

I wasn't entirely sure about the consequences of that, but I found https://github.com/Automattic/wp-calypso/pull/34994/commits/2971dc84f2c23ca50818e470a85e67f4a0f71612 which indicates the `QuerySitePlans` was added recently, for the new WordPress.com plan downgrade functionality in https://github.com/Automattic/wp-calypso/pull/34994.  I couldn't actually figure out how to reproduce a problem if it's not there, but reading that commit it seems like it's probably needed in general.  However, that also implies it's *not* needed for a disconnected Jetpack site.  So I guess it should be safe to just fix this to pass the selected site in when available, but not worry about the fact that it's unavailable for disconnected Jetpack sites.

chrismccluskey replied:
> I don't think the site property actually exists?

This confused me for a bit, because I don't randomly add in variables just to see what happens. I remember having access to this property, and React Dev Tools is showing it's an available property on that component, but where it comes from isn't obvious. If I'm reading the code correctly this property actually comes in [via the `ManagePurchase` component](https://github.com/Automattic/wp-calypso/blob/master/client/me/purchases/manage-purchase/index.jsx#L506) and is used as `selectedSite` anyway. 

Still, this doesn't need to be changed, so I have reverted it. I think the only reason it was changed in the first place was to allow the `ProductLink` component to display and was not reverted when it was clear the link itself wouldn't work because additional refactoring would need to be done on the components `ProductLink` links to. Given what you've pointed out with `QuerySitePlans`, I agree it's best not to disturb this component.

chrismccluskey commented:
> after a Jetpack site with a non-refundable purchase is disconnected, the "Cancel Subscription" link doesn't work... Maybe there is another server-side change that needs to be made (similar to the previous one for refundable purchases)?

Great catch, and that is exactly the case here. I've created D37026-code which gives the same treatment as the other endpoints to make this happen. I'm now able to cancel these subscriptions whereas previously I could not.

As for the other points you mentioned, I'm going to look into those now and see how best to handle. I feel like the code got switched on me a bit, as I swear some of those were coming in via grandparent components, but I can't see that's the case any longer.

chrismccluskey commented:
Ok, that's enough GitHub for me today. I've done a few things than anticipated on this PR, and hopefully didn't screw anything up or add a bunch more noise. I've addressed recent review comments from @DavidRothstein in the code, and think the code is likely in a good place for another (last?) review. However, I should also note a few other things that happened on this PR just in case it's a problem:

- Resolved conflicts in 6583ff4 in the GitHub web editor, accepting incoming changes, moving my change to the bottom of the incoming changes.
- Pulled to local but maybe did it incorrectly (used my IDE's GUI instead of CLI), committed changes from recent review comments, tried to push with CLI but failed, requiring a pull.
- Pulled from `origin`, which resulted in a merge commit (61e3e3d).
- Mistaking GitHub's "Resolve Conversation" button for phab's "Done" checkbox, I mistakenly started a review of.. that conversation? 🙃

DavidRothstein commented on client/me/purchases/cancel-purchase/button.jsx:
```diff
@@ -33,7 +33,7 @@ import { getDowngradePlanFromPurchase } from 'state/purchases/selectors';
 class CancelPurchaseButton extends Component {
 	static propTypes = {
 		purchase: PropTypes.object.isRequired,
-		selectedSite: PropTypes.object.isRequired,
```
`selectedSite` is still used in the component (although not required anymore) so shouldn't it still be declared here?

chrismccluskey replied:
You are correct. I mistakenly deleted these when I meant to only delete the `.isRequired` portion. Sorry, I'm new to propTypes. 🙃

DavidRothstein commented on client/me/purchases/cancel-purchase/index.jsx:
```diff
@@ -53,7 +52,6 @@ class CancelPurchase extends React.Component {
 		includedDomainPurchase: PropTypes.object,
 		purchase: PropTypes.object,
 		purchaseId: PropTypes.number.isRequired,
-		selectedSite: PropTypes.oneOfType( [ PropTypes.bool, PropTypes.object ] ),
```
This looks like it's still used one place in the class, so should it really be removed from the prop list?

chrismccluskey replied:
Same thing here.. removed the whole thing thinking that's what made it optional, didn't realize it was already optional because it lacked `.isRequired`.

DavidRothstein commented on client/me/purchases/cancel-purchase/index.jsx:
```diff
@@ -222,7 +215,7 @@ class CancelPurchase extends React.Component {
 						includedDomainPurchase={ this.props.includedDomainPurchase }
 						disabled={ this.state.cancelBundledDomain && ! this.state.confirmCancelBundledDomain }
-						selectedSite={ this.props.selectedSite }
+						siteSlug={ this.props.siteSlug }
```
It looks like `selectedSite` is still used by the `CancelPurchaseButton` component (when it's available) so don't we need to pass both `selectedSite` and `siteSlug` in here, rather than just the latter?

chrismccluskey replied:
Yep, you are correct. Thanks, I also missed that when refactoring back in `selectedSite`.

chrismccluskey commented:
Thanks again for your time and patience @DavidRothstein ! I've taken some additional time today to map and confirm every component relation to ensure the proper props have been properly propagated through to the right components. I've tested this again with a brand new account (on localhost, not calypso.live) and I'm able to cancel/refund disconnected Jetpack purchases successfully and also able to cancel/refund newly purchased wpcom subscriptions.

Aside from the seemingly unrelated, consistently failing tests of which I have no idea how to fix, I'm confident that there should be no remaining issues. 🤞

chrismccluskey commented:
Thanks for your patience, @DavidRothstein and I hope this is finally the last review. I've implemented your changes and everything works perfectly. The linter did force me to do a slight modification to your code but it wasn't a functional change (just had to move where `getSite` came from, as there was already a `from 'state/sites/selectors';`

chrismccluskey commented:
Just rebased from master. Let's see if that fixes the failing tests.

noahtallen commented:
Nice, looks like that fixed all the tests!

chrismccluskey commented:
@noahtallen yup! It's looking about ready to deploy now, after I do some testing to ensure I resolved merge conflicts correctly.
